doUpsert/index.js

```diff
@@ -56,24 +72,38 @@ module.exports = async function(context, jsonItem) {
           context,
           digitalTwin,
           jsonItem.$sourceId,
-          jsonItem.$relationshipId);
+          jsonItem.$relationshipId,
+      );
     } else { // relationship must be upserted
-      context.log.verbose(`upserting relationship ${jsonItem.$relationshipId}`);
-      await (async () => {
-        context.log.verbose('waiting 100ms');
-        sleep(100);
-        context.log.verbose('calling ADT relationship API');
-        await digitalTwin.upsertRelationship(
-            jsonItem.$sourceId,
-            jsonItem.$relationshipId,
-            jsonItem.relationship)
-            .catch((e) => {
-              context.log.error(`relationship ${jsonItem.$relationshipId}
+      try { // if relationship exists it must be updated
+        await digitalTwin.getRelationship(
```

Comment:
What's the purpose of this line? You try to get a relationship but you don't sotre/use the returned value?

Reply:
Basically testing the existence of the relationship but that's unclear. I'll do what you suggest in your other comments.

Reply:
Ok I remember now; `getDigitalTwin` and `getRelationship` both return an error if the twin or relationship does not exist:
```
[2021-12-17T10:53:23.808Z] Executed 'Functions.doUpsert' (Failed, Id=3333aa86-34de-43ca-a0ce-2605cee58a0e, Duration=1468ms)
[2021-12-17T10:53:23.810Z] System.Private.CoreLib: Exception while executing function: Functions.doUpsert. System.Private.CoreLib: Result: Failure
Exception: RestError: There is no digital twin instance that exists with the ID ThePrancingPony. Please verify that the twin id is valid and ensure that the twin is not deleted. See section on querying the twins https://aka.ms/adtv2query.
Stack: RestError: There is no digital twin instance that exists with the ID ThePrancingPony. Please verify that the twin id is valid and ensure that the twin is not deleted. See section on querying the twins https://aka.ms/adtv2query.
    at new RestError (C:\dev\azure-digital-twin-injector\node_modules\@azure\core-http\dist\index.js:2357:28)
    at handleErrorResponse (C:\dev\azure-digital-twin-injector\node_modules\@azure\core-http\dist\index.js:3208:17)
    at C:\dev\azure-digital-twin-injector\node_modules\@azure\core-http\dist\index.js:3147:18
    at processTicksAndRejections (node:internal/process/task_queues:96:5).
```
How can I test for the existence of the twin / relationship if I cannot store the result of the `getDigitalTwin` / `getRelationship` call?

Reply:
i checked the reference doc, there's not 'isExist' function. So i think you'll have to stick to your first implementation.
And to avoid too much exception hiding, just set the try/catch block on the get function.

Reply:
I narrowed the type of exception caught by the try/catch block to RestError to limit exception hiding.
I found some undesired behavior in my patch by doing so, I'm working on it right now. I doubt it'll be ready before the end of the day though. :/